### PR TITLE
assist: autoderef in generate delegate methods

### LINKED ISSUE #14438: Generate delegate methods... assist should look through deref
```rs
struct Strukt {
    field$0: Vec<u8>
}
```
invoking the assist here will not show delegation options for slice methods like `first` or `last` but it definitely should.

### PR DESCRIPTION
**Pull request by jhgg**
fixes #14438

### CODE REVIEW

File: crates/ide-assists/src/handlers/generate_delegate_methods.rs [hunk 1]
```diff
@@ -63,16 +65,23 @@ pub(crate) fn generate_delegate_methods(acc: &mut Assists, ctx: &AssistContext<'
...
+                    && f.is_visible_from(ctx.db(), current_module)
+                    && seen_names.insert(f.name(ctx.db()))
+                {
+                    methods.push(f)
+                }
             }
-        }
-        Option::<()>::None
-    });
+            Option::<()>::None
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_delegate_methods.rs:82]: Fun fact, `None::<()>` should also work!

**Review by Veykril — APPROVED** [review]

**Comment by Veykril** [comment]
Thanks!
@bors r+

**Comment by alibektas** [comment]
In a test case such as 
```rust
pub(crate) struct RibStack<R> {
    ribs: Vec<R>,
    used: usize,
}

impl<R: std::ops::Deref> std::ops::Deref for RibStack<R> {
    type Target = Vec<R>;

    fn deref(&self) -> &Self::Target {
        &self.ribs
    }
}
```

We can still select the assist option `Generate 'Deref' impl using 'ribs'` which results in ( conflicting ) duplicate implementation. 

```rust
pub(crate) struct RibStack<R> {
    ribs: Vec<R>,
    used: usize,
}

impl<R: std::ops::Deref> std::ops::Deref for RibStack<R> {
    type Target = Vec<R>;

    fn deref(&self) -> &Self::Target {
        &self.ribs
    }
}

impl<R: std::ops::Deref> std::ops::Deref for RibStack<R> {
    type Target = Vec<R>;

    fn deref(&self) -> &Self::Target {
        &self.ribs
    }
}
```

Is this really the desired behavior ?

**Comment by Veykril** [comment]
No that's a bug we should fix (though note this is a different assist than this PR is about), this pr is about the assist https://github.com/rust-lang/rust-analyzer/blob/master/crates/ide-assists/src/handlers/generate_delegate_methods.rs, your issue is about https://github.com/rust-lang/rust-analyzer/blob/master/crates/ide-assists/src/handlers/generate_deref.rs

**Pull request merged**